### ISSUE TITLE
request/response doesn't match

### ISSUE THREAD

**Issue by seven1240** [nats-io/nats.c#457]
Hi, first thanks the great software. We found a strange problem that we haven't figured out so far.

We use the C client `natsConnection_RequestString` to send a request to a Go lang service and wait for a response. sometimes we found that the response doesn't match the request.

We use JSON-RPC 2.0, so each request has a `id` in the JSON payload, and our logs in the Go side shows we published correct response to the request's INBOX and mirrors back the `id` in the json payload. But on the C side, our logs showed unmatched response.

```
uuid=914e81ba-68c1-4dfc-8668-e911522163a1 got invalid id=437e28b9-f5fd-49ce-9907-73a1559a63f3
uuid=1deb3d89-6d45-4fe5-a40a-a00a49837884 got invalid id=6eb567f8-5112-4fd2-9d1c-a0be767ad662
uuid=a52db4f3-155a-4f73-994b-bcf676f16a25 got invalid id=1deb3d89-6d45-4fe5-a40a-a00a49837884
uuid=9eecbfa9-a931-49d2-abdb-1e25c1911ff8 got invalid id=d2a6b4cf-a2fd-4040-8507-9a397c416561
uuid=e15dff9c-d506-41dc-8437-924e197de747 got invalid id=9eecbfa9-a931-49d2-abdb-1e25c1911ff8
uuid=4df52287-f371-4371-8217-3f26e9d1aeed got invalid id=e15dff9c-d506-41dc-8437-924e197de747
uuid=c4f1d207-984b-477a-ab7b-be925deb09b9 got invalid id=4df52287-f371-4371-8217-3f26e9d1aeed
```

so we only print the log when request/response `id` in the JSON payload is unmatched. from the pattern we can see that, from the last line, request `uuid=c4f1d207-984b-477a-ab7b-be925deb09b9` got a response contains an invalid `id=4df52287-f371-4371-8217-3f26e9d1aeed`, and that id was actually a request id in the line above it. That seems nats.c returned a response which belongs to another request.

So far we cannot reproduce this. We had seen this before in 2.3, but after we updated to 2.5 which didn't happen for a while. but yesterday we have one node(process) out of  4 (4 nodes are running the same code) hit that error again, and it been fixed after a restart.

We tried to trace that but so far haven't find the root cause. But before we dig more into the code we'd like to ask is there possibility that could lead the problem in the code?

I tried the unit test with 1000 threads and I cannot reproduce this. btw I changed the test code since the current test code send identical payloads on each request so it cannot detect a mismatched response, the following quick-and-dirty patch sends different payload on each request.

```c
diff --git a/test/test.c b/test/test.c
index 7b4da10..98b056f 100644
--- a/test/test.c
+++ b/test/test.c
@@ -4917,9 +4917,13 @@ _recvTestString(natsConnection *nc, natsSubscription *sub, natsMsg *msg,
         }
         case 4:
         {
-            arg->status = natsConnection_PublishString(nc,
+            arg->status = natsConnection_Publish(nc,
                                                        natsMsg_GetReply(msg),
-                                                       arg->string);
+                                                       natsMsg_GetData(msg),
+                                                       natsMsg_GetDataLength(msg));
+            // arg->status = natsConnection_PublishString(nc,
+            //                                            natsMsg_GetReply(msg),
+            //                                            arg->string);
             if (arg->status == NATS_OK)
                 arg->status = natsConnection_Flush(nc);
             break;
@@ -10846,7 +10850,8 @@ _sendRequest(void *closure)
 
     nats_Sleep(250);
 
-    s = natsConnection_RequestString(&msg, arg->nc, "foo", "Help!", 2000);
+    // s = natsConnection_RequestString(&msg, arg->nc, "foo", "Help!", 2000);
+    s = natsConnection_RequestString(&msg, arg->nc, "foo", arg->string, 2000);
     natsMutex_Lock(arg->m);
     if ((s == NATS_OK)
             && (msg != NULL)
@@ -10859,6 +10864,7 @@ _sendRequest(void *closure)
     else
     {
         arg->status = NATS_ERR;
+        printf("%s != %s\n", arg->string, natsMsg_GetData(msg));
     }
     natsMutex_Unlock(arg->m);
     natsMsg_Destroy(msg);
@@ -10870,11 +10876,13 @@ test_SimultaneousRequest(void)
     natsStatus          s;
      natsConnection      *nc       = NULL;
      natsSubscription    *sub      = NULL;
-     natsThread          *threads[10];
+     natsThread          *threads[1000];
      natsPid             serverPid = NATS_INVALID_PID;
      struct threadArg    arg;
      int                 i;
 
+    printf("version: %s\n", nats_GetVersion());
+
      s = _createDefaultThreadArgsForCbTests(&arg);
      if ( s != NATS_OK)
          FAIL("Unable to setup test!");
@@ -10883,24 +10891,40 @@ test_SimultaneousRequest(void)
      arg.status = NATS_OK;
      arg.control= 4;
 
-     serverPid = _startServer("nats://127.0.0.1:4222", NULL, true);
-     CHECK_SERVER_STARTED(serverPid);
+    //  serverPid = _startServer("nats://127.0.0.1:4222", NULL, true);
+    //  CHECK_SERVER_STARTED(serverPid);
 
-     s = natsConnection_ConnectTo(&nc, NATS_DEFAULT_URL);
+    //  s = natsConnection_ConnectTo(&nc, NATS_DEFAULT_URL);
+    //  s = natsConnection_ConnectTo(&nc, "nats://xswitch-nats:4222");
+     s = natsConnection_ConnectTo(&nc, "nats://192.168.7.7:4333");
      if (s == NATS_OK)
      {
          arg.nc = nc;
          s = natsConnection_Subscribe(&sub, nc, "foo", _recvTestString, (void*) &arg);
      }
 
-     for (i=0; i<10; i++)
+     for (i=0; i<1000; i++)
          threads[i] = NULL;
 
      test("Test simultaneous requests: ")
-     for (i=0; (s == NATS_OK) && (i<10); i++)
-         s = natsThread_Create(&(threads[i]), _sendRequest, (void*) &arg);
-
-     for (i=0; i<10; i++)
+     for (i=0; (s == NATS_OK) && (i<1000); i++)
+     {
+            struct threadArg *a = malloc(sizeof(struct threadArg));
+
+            s = _createDefaultThreadArgsForCbTests(a);
+            if ( s != NATS_OK)
+                FAIL("Unable to setup test!");
+            a->status = NATS_OK;
+            a->control= 4;
+            a->nc = nc;
+            a->m = arg.m;
+
+            char *str = malloc(20);
+            sprintf(str, "jj%d", i);
+            a->string = (const char *)str;
+         s = natsThread_Create(&(threads[i]), _sendRequest, (void*)a);
+     }
+     for (i=0; i<1000; i++)
      {
          if (threads[i] != NULL)
          {
@@ -10912,7 +10936,7 @@ test_SimultaneousRequest(void)
      natsMutex_Lock(arg.m);
      if ((s != NATS_OK)
              || ((s = arg.status) != NATS_OK)
-             || (arg.sum != 10))
+             || (arg.sum != 0))
      {
          s = NATS_ERR;
      }
```

I could submit a real pr to fix the test code if it looks good. And I'd like to provide more information to help trace this issue, either a lib issue or maybe our misuse of the code.

Thanks,
Seven.

**Comment by kozlovic** [comment]
The issue is fixed in https://github.com/nats-io/nats.c/pull/434 (which was avail in v2.5.1). Note that the fact that you mention that you have seen that pre v2.4.0 is worrisome, because this bug was introduced in v2.4.0 with PR #399.

This issue could not have happened pre v2.4.0, with that in mind, I would ask that you also check your producer code to see if there is any chance that the content of a request is incorrectly assigned to concurrent requests your code may perform.

As for updating the test: there is no need, this test was not meant to check the "unmatched" request/reply but just exercise possible races and handling of the internal pool of request infos. Running with 1000 threads would not work on Travis when running with valgrind (so if a change was made in this regard, the value would have to be conditionally set to lower in valgrind mode). But more importantly, as you said, it does not even help reproducing the issue with the "stealing" of responses.

**Comment by seven1240** [comment]
Thanks. we did use 2.3 on production for a long time and didn't find this problem. The problem we had maybe during we tried to upgrade into 2.4 which I cannot verify now.

As far as the test code I just changed to 1000 to see if it's easier to reproduce, not means it should be default in the code. And since the main focus is not the mismatch problem, you are right we don't need to change to test code to make it more complicated.

Thanks again, I'll test and get 2.5.1 into production soon and report back.
